README.md
<p align="center">
<img alt="Ant Cache logo" src="https://raw.githubusercontent.com/bonniss/ant-cache/main/logo.svg">
</p>

# Ant Cache

Simply fast in-memory key-value cache that supports TTL for Node.

## Getting started

```sh
npm install ant-cache
```

## Use cases

1. When you don't want (or need) to set up a separate service, like Redis or Memcached, for your local newspaper app.
2. When you just need a simple cache working like a mediator between your app and your real Backend.
3. When you just need a simple store for your config.

## Usage

With ES Module

 ```js
 import { AntCache } from 'ant-cache';
 ```

With CommonJS

 ```js
 const { AntCache } = require('ant-cache');
 ```

### Initialize

```js
// use default config
const cache = new AntCache();

// initialize with config
const cache = new AntCache({
  ttl: 120,  // in seconds
  checkPeriod: 10,  // in seconds
  maxKeys: 1000,  // could hold maximum 1000 key-value pairs
})
```

### Insert/Update

```js
// set a value with default ttl
cache.set('a number', 12);

// set a value with specific ttl 10 minutes
cache.set('a mighty commit', {
  id: '0efa4d37c3097bca9c58f4eaf75f86e7efdc518a',
  message: 'skyrocket'
}, 600);

// passing in an existing key to update value, `ttl` has no effect
cache.set('a mighty commit', {
  id: '0efa4d37c3097bca9c58f4eaf75f86e7efdc518a',
  message: 'enhanced skyrocket'
}, 600);

// if a value is inserted with `ttl` = 0, it will live permanently unless deleted manually
cache.set('permanent value', 'Linus Torvald', 0);
```

### Get

```js
// get a single key
cache.get('a number');

// get multiple keys, returns an object
cache.getMany('a mighty commit', 'permanent value');

// also accept array(s)
cache.getMany('a number', ['a mighty commit', 'permanent value']);
```

### Delete

```js
// delete a single key
cache.delete('a number');

// delete multiple keys, returns an object
cache.deleteMany('a mighty commit', 'permanent value');

// also accept array(s)
cache.deleteMany('a number', ['a mighty commit', 'permanent value']);

// clear cache
cache.flushAll();
```

### Stats

```js
cache.stats()
/*
{
  hits: 5,
  misses: 2,
  keys: 3
}
*/
```

### Hooks

By default, `deleteOnExpire` option is `true`, the cache will delete the expired value automatically. If you want to have your own logic, use `expired` hook:

```js
cache.on('expired', ({
  key,
  value,
  ttl,
  deleteCurrentKey,
}) => {
  console.info(`%s is expired`, key);
  deleteCurrentKey();
})
```

AntCache expose several hooks:

- before-delete
- after-delete
- before-set
- after-set
- expired

### Serialization

AntCache supports `serialize` and `deserialize` its content and TTLs. Any types supported by `superjson` should work as AntCache uses [superjson](https://github.com/blitz-js/superjson) under the hood.

The bigger the cache is, the longer it takes to complete. In worst cases, your app could be halted. So use those methods with caution.

```js
// stringify cache content and TTLs
const json = cache.serialize();

// parse JSON string and overwrite the cache content and TTLs.
cache.deserialize(json);
```

## API reference

Exhaustive API reference and examples:

[bonniss.github.io/ant-cache](https://bonniss.github.io/ant-cache/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.8](https://github.com/bonniss/ant-cache/compare/v0.0.7...v0.0.8) (2023-03-26)

### [0.0.7](https://github.com/bonniss/ant-cache/compare/v0.0.6...v0.0.7) (2023-03-26)

### [0.0.6](https://github.com/bonniss/ant-cache/compare/v0.0.5...v0.0.6) (2023-03-25)

### [0.0.5](https://github.com/bonniss/ant-cache/compare/v0.0.4...v0.0.5) (2023-03-25)

### [0.0.4](https://github.com/bonniss/ant-cache/compare/v0.0.3...v0.0.4) (2023-03-25)

### [0.0.3](https://github.com/bonniss/ant-cache/compare/v0.0.2...v0.0.3) (2023-03-24)

### 0.0.2 (2023-03-24)


src/lib/utils.spec.ts
import test from 'ava';

import { jsMapToObject, objectToJsMap } from './utils';

test('It should convert a map to an object', (t) => {
  const map = new Map<string, any>([
    ['input mobile matrix', 123],
    ['03730409f697554ae7eb1cab284b52845b6ca7a8', [2, 3, 4, 5]],
    [
      'Account',
      {
        en: 'Churchill',
      },
    ],
  ]);
  const obj = jsMapToObject(map);

  t.deepEqual(obj, {
    'input mobile matrix': 123,
    '03730409f697554ae7eb1cab284b52845b6ca7a8': [2, 3, 4, 5],
    Account: {
      en: 'Churchill',
    },
  });
});

test('It should convert an object to a map', (t) => {
  const obj = {
    'input mobile matrix': 123,
    '03730409f697554ae7eb1cab284b52845b6ca7a8': [2, 3, 4, 5],
    Account: {
      en: 'Churchill',
    },
  };
  const map = objectToJsMap(obj);

  t.deepEqual(
    map,
    new Map<string, any>([
      ['input mobile matrix', 123],
      ['03730409f697554ae7eb1cab284b52845b6ca7a8', [2, 3, 4, 5]],
      [
        'Account',
        {
          en: 'Churchill',
        },
      ],
    ])
  );
});

test("It should overwrite target map with object's values", (t) => {
  const obj = {
    'input mobile matrix': 123,
    '03730409f697554ae7eb1cab284b52845b6ca7a8': [2, 3, 4, 5],
    Account: {
      en: 'Churchill',
    },
  };

  const map: Map<any, any> = new Map([
    [
      'Account',
      {
        ja: 'Kozumi',
      },
    ],
  ]);

  const modifiedMap = objectToJsMap(obj, map);
  // same reference
  t.is(map === modifiedMap, true);

  t.deepEqual(map.get('Account'), {
    en: 'Churchill',
  });
});


src/lib/ant-cache.spec.ts
import test from 'ava';

import {
  AntCacheConfig,
  AntCacheValue,
  OnExpireCallbackInput,
} from '../types/ant-cache';

import { AntCache } from './ant-cache';
import { MaxKeysExceedError } from './max-keys-exceed-error';

const defaultConfig: AntCacheConfig = {
  ttl: 4,
  checkPeriod: 1,
  maxKeys: 0,
  deleteOnExpire: true,
};

const setup = (config?: AntCacheConfig) => {
  const merged = { ...defaultConfig, ...config };
  return new AntCache(merged);
};

const ONE_SECOND_IN_MILLISECS = 1300;
const pSetTimeout = (callback: () => void, timeoutInSeconds: number) =>
  new Promise<void>((resolve) => {
    setTimeout(() => {
      callback();
      void resolve();
    }, timeoutInSeconds * ONE_SECOND_IN_MILLISECS);
  });

const genKey = (): string =>
  [...Array(30)].map(() => Math.random().toString(36)[2]).join('');
const floodByPrimitives = (cache: AntCache): [string[], any[]] => {
  const values = [
    null,
    undefined,
    // string
    'Liverpool',
    // number
    NaN,
    Number.MAX_VALUE,
    Infinity,
    -Infinity,
    // boolean
    true,
    false,
    // bigint
    BigInt(Number.MAX_SAFE_INTEGER) + BigInt(1),
  ];
  const keys: string[] = values.map((_, index) => `key${index.toString()}`);
  keys.forEach((key, index) => {
    const value = values[index];
    cache.set(key, value);
  });
  return [keys, values];
};
const floodByObjects = (cache: AntCache): [string[], any[]] => {
  const values = [
    Object.create(null),
    Object.create({}),
    // Object literals
    { name: 'Liverpool' },
    // bigint
    [],
    [1, 'string', null, true, BigInt(1)],
  ];
  const keys: string[] = values.map((_, index) => `key${index.toString()}`);
  keys.forEach((key, index) => {
    const value = values[index];
    cache.set(key, value);
  });
  return [keys, values];
};

/**
 * Init & Config
 */
test('It should be initialized', async (t) => {
  const cache = setup();
  t.truthy(cache);
});

test('It should be configured', (t) => {
  const cache = setup();
  t.is(cache.config.ttl, defaultConfig.ttl);
  t.is(cache.config.checkPeriod, defaultConfig.checkPeriod);
  t.is(cache.config.maxKeys, defaultConfig.maxKeys);
  t.is(cache.config.deleteOnExpire, true);
});

/**
 * Get/Set
 */
test('It should be able to set primitive values', (t) => {
  const cache = setup();
  const [keys] = floodByPrimitives(cache);
  for (const key of keys) {
    t.is(cache.mainCache.has(key), true);
  }
});

test('It should has the key after set', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(key, key);
  t.is(cache.has(key), true);
});

test('It should be able to get primitive values', (t) => {
  const cache = setup();
  const [keys, values] = floodByPrimitives(cache);
  keys.forEach((key, index) => {
    const value = values[index];
    t.is(cache.get(key), value);
  });
});

test('It should be able to get a group of keys', (t) => {
  const cache = setup();
  const [keys] = floodByPrimitives(cache);
  const groupSize = Math.ceil(keys.length / 2);
  const keyGroup = keys.slice(0, groupSize);

  const obj = cache.getMany(...keyGroup);

  for (const key of keyGroup) {
    t.is(obj[key], cache.get(key));
  }
});

test('If a key not exist, it should returns `undefined`', (t) => {
  const cache = setup();
  const randomKey = genKey();
  t.is(cache.get(randomKey), undefined);
});

test('It should be able to set objects', (t) => {
  const cache = setup();
  const [keys] = floodByObjects(cache);
  for (const key of keys) {
    t.is(cache.mainCache.has(key), true);
  }
});

test('It should be able to get objects', (t) => {
  const cache = setup();
  const [keys, values] = floodByObjects(cache);
  keys.forEach((key, index) => {
    const value = values[index];
    t.deepEqual(cache.get(key), value);
  });
});

/**
 * Keys
 */
test('It should returns all keys', (t) => {
  const cache = setup();
  const [keys] = floodByPrimitives(cache);
  t.deepEqual(cache.keys(), keys);
});

test('It should throws error when exceed `maxKeys`', (t) => {
  const maxKeys = 2;
  const cache = setup({
    maxKeys,
  });
  const fn = () => {
    for (let index = 0; index < maxKeys + 1; index++) {
      cache.set(genKey(), index);
    }
  };

  const error = t.throws(fn, { instanceOf: MaxKeysExceedError });

  t.log(error);
});

/**
 * Stats
 */
test('It should record stats', (t) => {
  const cache = setup();
  const expectedHits = 10;
  for (let index = 0; index < expectedHits; index++) {
    const key = genKey();
    cache.set(key, key);
    // increase hits
    cache.get(key);
  }

  const expectedMisses = 10;
  for (let index = 0; index < expectedMisses; index++) {
    // increase misses
    cache.get(genKey());
  }

  t.deepEqual(cache.stats(), {
    size: expectedHits,
    hits: expectedHits,
    misses: expectedMisses,
  });
});

/**
 * Delete
 */
test('It should delete an existing key', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(key, key);
  cache.delete(key);
  t.is(cache.get(key), undefined);
});

test('It should be able to delete multiple keys at once', (t) => {
  const cache = setup();
  const [keys] = floodByPrimitives(cache);
  const groupSize = Math.ceil(keys.length / 2);
  const keyGroup = keys.slice(0, groupSize);

  cache.deleteMany(...keyGroup);
  t.is(cache.size(), keys.length - groupSize);
});

test('It should clear all', (t) => {
  const cache = setup();
  floodByPrimitives(cache);
  cache.flushAll();
  t.is(cache.keys.length, 0);
});

/**
 * TTL
 */
test('Value should be deleted when expires by default ttl', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(
    key,
    `this value should be expired after ${defaultConfig.ttl} secs`
  );
  return pSetTimeout(() => {
    t.is(cache.get(key), undefined);
  }, defaultConfig.ttl);
});

test('Value should be deleted when expires by custom ttl', (t) => {
  const cache = setup();
  const ttl = 6;
  const key = genKey();
  cache.set(key, `this value should be expired after ${ttl} secs`, ttl);
  return pSetTimeout(() => {
    t.is(cache.get(key), undefined);
  }, ttl);
});

test('Value should exists when not expires', (t) => {
  const cache = setup();
  const timeout = defaultConfig.ttl / 2;
  const key = genKey();
  cache.set(key, `this value should exists after ${timeout} secs`);
  return pSetTimeout(() => {
    t.truthy(cache.get(key));
  }, timeout);
});

test('It should work as an enhanced JS Map if checkPeriod = 0', (t) => {
  const cache = setup({
    checkPeriod: 0,
    ttl: 2,
  });
  const key = genKey();
  cache.set(key, key);

  return pSetTimeout(() => {
    t.is(cache.get(key), key);
  }, 2 * cache.config.ttl);
});

test('It should not delete expired keys automatically when `deleteOnExpire` = false', (t) => {
  const cache = setup({
    ttl: 2,
    deleteOnExpire: false,
  });
  const key = genKey();
  cache.set(key, key);
  return pSetTimeout(() => {
    t.is(cache.get(key), key);
  }, 2 * cache.config.ttl);
});

test('It should not delete expired keys automatically if the key is inserted with `ttl` = 0', (t) => {
  const cache = setup({
    ttl: 2,
  });
  const key = genKey();
  cache.set(key, key, 0);
  return pSetTimeout(() => {
    t.is(cache.get(key), key);
  }, 2 * cache.config.ttl);
});

// Hooks
test('`before-set` and `after-set` hooks should be called', (t) => {
  const cache = setup();
  const key = genKey();
  cache.on('before-set', () => {
    t.is(cache.get(key), undefined);
  });
  cache.on('after-set', () => {
    t.is(cache.get(key), key);
  });
  cache.set(key, key);
});

test('`before-delete` and `after-delete` hooks should be called', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(key, key);
  cache.on('before-delete', () => {
    t.is(cache.get(key), key);
  });
  cache.on('after-delete', () => {
    t.is(cache.get(key), undefined);
  });
  cache.delete(key);
});

test('`expired` hook should be called and expired key should be delete after `deleteCurrentKey` invoked', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(key, key);
  cache.on('expired', (inp: OnExpireCallbackInput) => {
    t.log(
      `{ key: ${inp.key}, value: ${inp.value}, ttl: ${inp.ttl} } has expired`
    );
    t.is(inp.key, key);
    t.is(inp.value, key);
    inp.deleteCurrentKey();
  });

  return pSetTimeout(() => {
    t.is(cache.get(key), undefined);
  }, defaultConfig.ttl);
});

test('`expired` hook should be called and expired key should be intact when `deleteCurrentKey` not invoked', (t) => {
  const cache = setup();
  const key = genKey();
  cache.set(key, key);
  cache.on('expired', (inp: OnExpireCallbackInput) => {
    t.log(
      `{ key: ${inp.key}, value: ${inp.value}, ttl: ${inp.ttl} } has expired`
    );
    t.is(inp.key, key);
    t.is(inp.value, key);
  });

  return pSetTimeout(() => {
    t.is(cache.get(key), key);
  }, defaultConfig.ttl);
});

/**
 * Serialization
 */
const floodCacheRealScenario = (cache: AntCache) => {
  cache.set('date', new Date());
  cache.set('bigint', BigInt(1000));
  cache.set('number', 1000);
  cache.set('null', null);
  cache.set('string', '738875b4b2c1c70eff4666a88bd1fc1c9b816d99');
  cache.set('array', ['Jersey Total', null, 123]);
  cache.set(
    'set',
    new Set([
      1,
      null,
      5,
      {
        location: 'Peru',
      },
    ])
  );
  cache.set(
    'map',
    new Map<string, AntCacheValue>([
      ['date', new Date()],
      ['bigint', BigInt(1000)],
      ['number', 1000],
      ['null', null],
      ['string', '738875b4b2c1c70eff4666a88bd1fc1c9b816d99'],
    ])
  );
};

test('It should be serializable', (t) => {
  const cache = setup();
  floodCacheRealScenario(cache);
  t.notThrows(() => {
    const json = cache.serialize();
    t.log(json);
    t.truthy(json);
  });
});

test('It should be deserializable and recover', (t) => {
  const cache = setup();
  floodCacheRealScenario(cache);
  const json = cache.serialize();
  const content = cache.getAll();
  cache.flushAll();
  cache.deserialize(json);
  const newContent = cache.getAll();
  t.deepEqual(newContent, content);
});

test('It should be deserializable and overwritten', (t) => {
  const cache = setup({
    checkPeriod: 1,
    ttl: 2,
  });
  floodCacheRealScenario(cache);
  const oldValue = cache.get('string');
  const json = cache.serialize();
  cache.flushAll();
  cache.set('string', 'Liverpool', 6);
  cache.deserialize(json);
  t.is(cache.get('string'), oldValue);
  return pSetTimeout(() => {
    t.is(cache.get('string'), undefined);
  }, 3);
});


src/types/config.d.ts
/**
 * Ant Cache Configuration
 */
interface AntCacheConfig {
  /**
   * The period in seconds, as a number, used for the automatic delete check interval.
   * 0 = no periodic check.
   *
   * Default: `15 seconds`
   */
  readonly checkPeriod?: number;

  /**
   * Default TTL in seconds.
   *
   * Default: double the default `checkPeriod`
   */
  readonly ttl?: number;

  /**
   * Specifies a maximum amount of keys that can be stored in the cache. If a new item is set and the cache is full, an error is thrown and the key will not be saved in the cache.
   *
   * `0` disables the key limit.
   *
   * Default: `0`
   */
  readonly maxKeys?: number;

  /**
   * whether keys will be deleted automatically when they expire.
   */
  readonly deleteOnExpire?: boolean;
}

export default AntCacheConfig;


src/types/ant-cache.d.ts
/**
 * Event name
 */
export type AntCacheEvent =
  | 'expired'
  | 'before-set'
  | 'after-set'
  | 'before-delete'
  | 'after-delete';

/**
 * Value type to store
 *
 * Cover native JSON types, plus Set, Map, Date, bigint
 */
export type AntCacheValue =
  | number
  | string
  | boolean
  | null
  | bigint
  | Date
  | Map<AntCacheValue, AntCacheValue>
  | Set<AntCacheValue>
  | { [key: string]: AntCacheValue }
  | AntCacheValue[];

export type AntCacheStats = {
  size: number;
  hits: number;
  misses: number;
};

export type OnExpireCallbackInput = {
  key: string;
  value: AntCacheValue;
  ttl: number;
  deleteCurrentKey: () => void;
};

export interface AntCacheConfig {
  /**
   * The period in seconds, as a number, used for the automatic delete check interval.
   * 0 = no periodic check.
   *
   * @default: 30
   */
  readonly checkPeriod?: number;

  /**
   * Default TTL in seconds.
   *
   * @default: double the default `checkPeriod`
   */
  readonly ttl?: number;

  /**
   * Specifies a maximum amount of keys that can be stored in the cache. If a new item is set and the cache is full, an error is thrown and the key will not be saved in the cache.
   *
   * `0` disables the key limit.
   *
   * @default: 0
   */
  readonly maxKeys?: number;

  /**
   * whether keys will be deleted automatically when they expire.
   */
  readonly deleteOnExpire?: boolean;
}

export as namespace AntCache;


src/lib/ant-cache.ts
import { EventEmitter } from 'events';

import superjson from 'superjson';

import {
  AntCacheConfig,
  AntCacheEvent,
  AntCacheStats,
  AntCacheValue,
  OnExpireCallbackInput,
} from '../types/ant-cache';

import { MaxKeysExceedError } from './max-keys-exceed-error';
import { jsMapToObject, objectToJsMap } from './utils';

/**
 * Event emitter to listen to specific hooks
 */
export class AntCacheEventEmitter extends EventEmitter {}

const DEFAULT_CHECK_PERIOD = 30;

/**
 * Default cache configuration
 */
export const defaultAntCacheConfig: AntCacheConfig = {
  checkPeriod: DEFAULT_CHECK_PERIOD,
  ttl: 2 * DEFAULT_CHECK_PERIOD,
  maxKeys: 0,
  deleteOnExpire: true,
};

/**
 * ## Get stated
 *
 * ### ES Module
 *
 * ```js
 * import { AntCache } from 'ant-cache';
 * ```
 *
 * ### Common JS
 *
 * ```js
 * const { AntCache } = require('ant-cache');
 * ```
 *
 * ## Usage
 *
 * ### Use cache with TTL
 *
 * ```js
 * // initialize using default config
 * const cache = new AntCache();
 *
 * // initialize with config
 * const cache = new AntCache({
 *   ttl: 120,  // in seconds
 *   checkPeriod: 10,  // in seconds
 *   maxKeys: 1000,  // could hold maximum 1000 key-value pairs
 * })
 * ```
 *
 * ### Use as an enhanced JS `Map` (no TTL)
 *
 * ```js
 * const cache = new AntCache({
 *   checkPeriod: 0
 * });
 * ```
 */
export class AntCache {
  /**
   * The config of the cache
   *
   * Initialized once in the constructor, then read-only
   */
  config: AntCacheConfig;

  /**
   * The main store that interacts with the outside world.
   *
   * Methods to get the cache info, like `size` and `keys`, apply on this.
   */
  mainCache: Map<string, AntCacheValue>;

  /**
   * Store the created timestamp of every key.
   *
   * Must be synced with the main cache every time a new value is set.
   *
   */
  private _createdTsMap: Map<string, number>;

  /**
   * Store the ttl of every key.
   *
   * Must be synced with the main cache every time a new value is set.
   *
   */
  private _ttlMap: Map<string, number>;

  /**
   * The interval timer
   */
  private _timerId: ReturnType<typeof setInterval>;

  /**
   * The event emitter
   */
  private _emitter: AntCacheEventEmitter;

  /**
   * Cache hits
   */
  private _hits = 0;

  /**
   * Cache misses
   */
  private _misses = 0;

  /**
   * `true` if `config.checkPeriod` = 0
   */
  private _checkPeriodDisabled = false;

  /**
   * @param config if no arg passed in, use defaultAntCacheConfig
   *
   * ```ts
   * const DEFAULT_CHECK_PERIOD = 30;
   * export const defaultAntCacheConfig: AntCacheConfig = {
   *    checkPeriod: DEFAULT_CHECK_PERIOD,
   *    ttl: 2 * DEFAULT_CHECK_PERIOD,
   *    maxKeys: 0,
   * };
   * ```
   */
  constructor(config?: AntCacheConfig) {
    this.config = { ...defaultAntCacheConfig, ...config };

    this.mainCache = new Map();
    this._createdTsMap = new Map();
    this._ttlMap = new Map();
    this._emitter = new AntCacheEventEmitter();

    if (this.config.checkPeriod > 0) {
      this._timerId = setInterval(
        this.handleExpired.bind(this),
        this.config.checkPeriod * 1000
      );
    } else {
      this._checkPeriodDisabled = true;
    }
  }

  /**
   * This method is called every `checkPeriod` second interval
   * to find and handle expired keys.
   */
  private handleExpired() {
    // check expired keys
    this._ttlMap.forEach((ttlInMillisecs, key) => {
      const ts = this._createdTsMap.get(key);
      const isExpired = ts !== undefined && +new Date() - ts > ttlInMillisecs;
      if (isExpired) {
        if (this._emitter.eventNames().includes('expired' as AntCacheEvent)) {
          const value = this.mainCache.get(key);
          this._emitter.emit(
            'expired' as AntCacheEvent,
            {
              key,
              value,
              ttl: ttlInMillisecs / 1000,
              deleteCurrentKey: () => this._del(key),
            } as OnExpireCallbackInput
          );
        } else {
          this.config.deleteOnExpire && this.delete(key);
        }
      }
    });
  }

  /**
   * Insert or update a key.
   *
   * ```js
   * const cache = new AntCache({ ttl: 4, checkPeriod: 1 })
   *
   * // use default TTL
   * cache.set('default ttl', 'live for 4 seconds');
   *
   * // use custom TTL
   * cache.set('custom ttl', 'live for 10 seconds', 10);
   *
   * // live permanently
   * cache.set('no ttl', 'this will live forever unless be deleted manually', 0);
   * ```
   *
   * @param key should be `string`
   * @param val
   * @param ttl in seconds, only take effect when inserting a new key. If `ttl` = 0, there is no TTL at all
   */
  public set(key: string, val: AntCacheValue, ttl = this.config.ttl) {
    const maxKeys = this.config.maxKeys;
    if (maxKeys && this.mainCache.size === maxKeys) {
      throw new MaxKeysExceedError(maxKeys);
    }
    this._emitter.emit('before-set' as AntCacheEvent, key, val);
    const isNewKey = !this.mainCache.has(key);
    this.mainCache.set(key, val);
    if (!this._checkPeriodDisabled && isNewKey && ttl) {
      this._ttlMap.set(key, ttl * 1000);
      this._createdTsMap.set(key, +new Date());
    }
    this._emitter.emit('after-set' as AntCacheEvent, key, val);
  }

  /**
   * Retrieve value for `key`
   *
   * @param key
   * @returns `undefined` if `key` not found
   */
  public get(key: string) {
    if (this.mainCache.has(key)) {
      this._hits++;
      return this.mainCache.get(key);
    }

    this._misses++;
    return undefined;
  }

  /**
   * Retrieve cache stats
   */
  public stats(): AntCacheStats {
    return {
      size: this.mainCache.size,
      hits: this._hits,
      misses: this._misses,
    };
  }

  /**
   * Retrieve values of key groups
   *
   * @param key
   * @returns an object contains values of `keys`
   */
  public getMany(...args: (string | string[])[]) {
    const keys = args.flat();
    const results: Record<string, AntCacheValue> = {};

    for (const key of keys) {
      const val = this.mainCache.get(key);
      results[key] = val;
    }

    return results;
  }

  /**
   * Retrieve values of all keys
   *
   * @returns an object of { [key]: value }
   */
  public getAll() {
    return jsMapToObject<AntCacheValue>(this.mainCache);
  }

  /**
   * Get all keys
   *
   * @returns an array contains all keys
   */
  public keys() {
    return Array.from(this.mainCache.keys());
  }

  /**
   * Get all values
   *
   * @returns an array contains all keys
   */
  public values() {
    return Array.from(this.mainCache.values());
  }

  /**
   * Get cache size
   *
   * @returns the size of the cache
   */
  public size() {
    return this.mainCache.size;
  }

  /**
   * Check if the cache contains `key`
   *
   * @param key
   * @returns `true` if `key` in the cache, otherwise `false`
   */
  public has(key: string) {
    return this.mainCache.has(key);
  }

  /**
   * Delete a single `key`
   *
   * Note: this method emits 'before-delete' hook,
   * then deletes the `key`,
   * then emits 'after-delete' hook.
   */
  public delete(key: string) {
    this._emitter.emit('before-delete' as AntCacheEvent, key);
    this._del(key);
    this._emitter.emit('after-delete' as AntCacheEvent, key);
  }

  /**
   * Delete multiple keys
   *
   * @param keys an array of keys to delete
   */
  public deleteMany(...args: (string | string[])[]) {
    const keys = args.flat();
    for (const key of keys) {
      this.delete(key);
    }
  }

  /**
   * Delete a key from maps in sync.
   *
   * For internal use only.
   */
  private _del(key: string) {
    this.mainCache.delete(key);
    this._createdTsMap.delete(key);
    this._ttlMap.delete(key);
  }

  /**
   * Empty the cache
   */
  public flushAll() {
    this.mainCache.clear();
    this._createdTsMap.clear();
    this._ttlMap.clear();
  }

  /**
   * Stringify the cache content and ttl.
   *
   * _Use with caution as the bigger the cache is, the longer it takes to stringify._
   *
   * Leave the logic handling long-running task up to you.
   *
   * @returns JSON string to be consumed in `deserialize`
   */
  public serialize(): string {
    const content = this.getAll();
    const ttls = jsMapToObject<number>(this._ttlMap);
    return superjson.stringify([content, ttls]);
  }

  /**
   * Parse the input JSON string and attempt to upsert into the current cache.
   *
   * __JSON string should be the output of method `serialize`__.
   *
   * _Existing keys will be overwritten both value and TTL_. Created timestamps will be reset.
   *
   * _Use with caution as the bigger the string is, the longer it takes to parse._
   *
   * Leave logic handling long-running task up to you.
   *
   * @returns
   */
  public deserialize(json: string) {
    const [content, ttls]: [
      Record<string, AntCacheValue>,
      Record<string, number>
    ] = superjson.parse(json);

    void objectToJsMap(content, this.mainCache);
    void objectToJsMap(ttls, this._ttlMap);

    for (const key in ttls) {
      this._createdTsMap.set(key, +new Date());
    }
  }

  /**
   * Adds a listener at the end of the listeners array for the specified event
   *
   * @param eventName an event name comply to `AntCacheEvent`
   * @param callback a callback that will be call when the event emitted
   */
  public on(eventName: AntCacheEvent, callback: (...args: any[]) => void) {
    this._emitter.on(eventName, callback);
  }

  /**
   * Prepare to close the cache.
   *
   * Dispose the timer and the remove all event listeners.
   */
  public dispose() {
    clearInterval(this._timerId);
    this._emitter.removeAllListeners();
  }

  /**
   * Alias for `dispose`
   */
  public readonly close = this.dispose;
}

export { MaxKeysExceedError };
export type {
  AntCacheConfig,
  AntCacheEvent,
  AntCacheStats,
  AntCacheValue,
  OnExpireCallbackInput,
};
